osquery/tables/system/windows/shellbags.cpp

```diff
@@ -0,0 +1,441 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <osquery/core/core.h>
+#include <osquery/core/tables.h>
+#include <osquery/logger/logger.h>
+#include <osquery/tables/system/windows/registry.h>
+#include <osquery/utils/conversions/windows/windows_time.h>
+#include <osquery/utils/windows/shellitem.h>
+
+#include <string>
+#include <vector>
+
+namespace osquery {
+namespace tables {
+constexpr auto kShellBagPath =
+    "\\Local Settings\\Software\\Microsoft\\Windows\\Shell\\BagMRU";
+constexpr auto kShellBagPathNtuser =
+    "\\Software\\Microsoft\\Windows\\Shell\\BagMRU";
+
+std::string buildPath(const std::vector<std::string>& build_shellbag) {
+  std::string full_path = "";
+  for (const auto& path : build_shellbag) {
+    full_path += path;
+  }
+  return full_path;
+}
+
+std::string guidLookup(const std::string& guid) {
+  QueryData guid_data;
+  queryKey("HKEY_LOCAL_MACHINE\\SOFTWARE\\Classes\\CLSID\\{" + guid + "}",
+           guid_data);
+  for (const auto& rKey : guid_data) {
+    auto key_type = rKey.find("type");
+    auto key_path = rKey.find("path");
+    if (key_type == rKey.end() || key_path == rKey.end()) {
+      continue;
+    }
+    auto key_name = rKey.at("name");
+    if (key_name == "(Default)") {
+      if (rKey.at("data") == "") {
+        return "{" + guid + "}";
+      }
+      return rKey.at("data");
+    }
+  }
+  return "{" + guid + "}";
+}
+
+void parseShellData(const std::string& shell_data,
+                    std::vector<std::string>& build_shellbag,
+                    QueryData& results,
+                    const std::string& sid,
+                    const std::string& source) {
```

Comment:
This encapsulates a ton of complex logic and doesn't seem to call out to anything else (I may have missed that). Can you put some unit tests in for it?